vault/identity_store_oidc.go

```diff
@@ -0,0 +1,1349 @@
+	signer, err := jose.NewSigner(signingKey, &jose.SignerOptions{})
+	if err != nil {
+		return "", err
+	}
+
+	signature, err := signer.Sign(payload)
+	if err != nil {
+		return "", err
+	}
+
+	signedIdToken, err := signature.CompactSerialize()
+	if err != nil {
+		return "", err
+	}
+
+	return signedIdToken, nil
+}
+
+func (i *IdentityStore) pathOIDCRoleExistenceCheck(ctx context.Context, req *logical.Request, d *framework.FieldData) (bool, error) {
+	role, err := i.getOIDCRole(ctx, req.Storage, d.Get("name").(string))
+	if err != nil {
+		return false, err
+	}
+
+	return role != nil, nil
+}
+
+// handleOIDCCreateRole is used to create a new role or update an existing one
+func (i *IdentityStore) pathOIDCCreateUpdateRole(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	name := d.Get("name").(string)
+
+	var role role
+	entry, err := req.Storage.Get(ctx, roleConfigPath+name)
+	if err != nil {
+		return nil, err
+	}
+	if entry != nil {
+		if err := entry.DecodeJSON(&role); err != nil {
+			return nil, err
+		}
+	}
+
+	if key, ok := d.GetOk("key"); ok {
+		role.Key = key.(string)
+	} else if req.Operation == logical.CreateOperation {
+		role.Key = d.Get("key").(string)
+	}
+
+	if role.Key == "" {
+		return logical.ErrorResponse("key must be provided"), nil
+	}
+
+	// validate that key exists
+	entry, err = req.Storage.Get(ctx, namedKeyConfigPath+role.Key)
+	if err != nil {
+		return nil, err
+	}
+	if entry == nil {
+		return logical.ErrorResponse("Key %q does not exist", role.Key), nil
+	}
+
+	if template, ok := d.GetOk("template"); ok {
+		role.Template = template.(string)
+	} else if req.Operation == logical.CreateOperation {
+		role.Template = d.Get("template").(string)
+	}
+
+	// Validate that template can be parsed and results in valid JSON
+	if role.Template != "" {
+		_, populatedTemplate, err := identity.PopulateString(&identity.PopulateStringInput{
+			Mode:   identity.JSONTemplating,
+			String: role.Template,
+			Entity: new(identity.Entity),
+			Groups: make([]*identity.Group, 0),
+			// namespace?
+		})
+
+		if err != nil {
+			return logical.ErrorResponse("Error parsing template: %s", err.Error()), nil
+		}
+
+		var tmp map[string]interface{}
+		if err := json.Unmarshal([]byte(populatedTemplate), &tmp); err != nil {
+			return logical.ErrorResponse("Error parsing template JSON: %s", err.Error()), nil
+		}
+
+		for key := range tmp {
+			if strutil.StrListContains(requiredClaims, key) {
+				return logical.ErrorResponse(`Top level key %q not allowed. Restricted keys: %s`,
+					key, strings.Join(requiredClaims, ", ")), nil
+			}
+		}
+	}
+
+	if ttl, ok := d.GetOk("ttl"); ok {
+		role.TokenTTL = time.Duration(ttl.(int)) * time.Second
+	} else if req.Operation == logical.CreateOperation {
+		role.TokenTTL = time.Duration(d.Get("ttl").(int)) * time.Second
+	}
+
+	// create role path
+	if role.RoleID == "" {
+		roleID, err := uuid.GenerateUUID()
```

Comment:
Should we use a base62 ID? Any reason for it to be UUID formatted?

Reply:
base62 is fine. Changed to a 26-char one like elsewhere in Vault.